primitives/npos-elections/src/phragmms.rs

```diff
@@ -0,0 +1,382 @@
+/// Returns a new `CandidatePtr` to the winner candidate. The score of the candidate is updated and
+/// can be read from the returned pointer.
+///
+/// This is an internal part of the [`balanced_heuristic`].
+pub(crate) fn calculate_max_score<AccountId: IdentifierT, P: PerThing>(
+	candidates: &[CandidatePtr<AccountId>],
+	voters: &[Voter<AccountId>],
+) -> CandidatePtr<AccountId> where ExtendedBalance: From<InnerOf<P>> {
+	for c_ptr in candidates.iter() {
+		let mut candidate = c_ptr.borrow_mut();
+		if !candidate.elected {
+			candidate.score = Rational128::from(1, P::ACCURACY.into());
+		}
+	}
+
+	for voter in voters.iter() {
+		let mut denominator_contribution: ExtendedBalance = 0;
+
+		// gather contribution from all elected edges.
+		for edge in voter.edges.iter() {
+			let edge_candidate = edge.candidate.borrow();
+			if edge_candidate.elected {
+				let edge_contribution: ExtendedBalance = P::from_rational_approximation(
+					edge.weight,
+					edge_candidate.backed_stake,
+				).deconstruct().into();
+				denominator_contribution += edge_contribution;
+			}
+		}
+
+		// distribute to all _unelected_ edges.
+		for edge in voter.edges.iter() {
+			let mut edge_candidate = edge.candidate.borrow_mut();
+			if !edge_candidate.elected {
+				let prev_d = edge_candidate.score.d();
+				edge_candidate.score = Rational128::from(1, denominator_contribution + prev_d);
+			}
+		}
+	}
+
+	// finalise the score value, and find the best.
+	let mut best_score = Rational128::zero();
+	// TODO: this function should be able to return Option.
+	let mut best_candidate = Rc::clone(&candidates[0]);
+	for c_ptr in candidates.iter() {
+		let mut candidate = c_ptr.borrow_mut();
+		if candidate.approval_stake > 0  {
+			// finalise the score value.
+			let score_d = candidate.score.d();
+			let one: ExtendedBalance = P::ACCURACY.into();
+			// TODO: the accuracy here is questionable.
+			let score_n = candidate.approval_stake.checked_mul(one).unwrap_or_else(|| Bounded::max_value());
+			candidate.score = Rational128::from(score_n, score_d);
+
+			// check if we have a new winner.
+			if !candidate.elected && candidate.score > best_score {
+				best_score = candidate.score;
+				best_candidate = Rc::clone(&c_ptr);
+			}
+		} else {
+			candidate.score = Rational128::zero();
+		}
+	}
+
+	best_candidate
+}
+
+/// Update the weights of `voters` given that `elected_ptr` has been elected in the previous round.
+///
+/// Updates `voters` in place.
+///
+/// This is an internal part of the [`balanced_heuristic`] and should be called after
+/// [`calculate_max_score`].
+pub(crate) fn apply_elected<AccountId: IdentifierT>(
+	voters: &mut Vec<Voter<AccountId>>,
+	elected_ptr: CandidatePtr<AccountId>,
+) {
+	let mut elected = elected_ptr.borrow_mut();
+	let cutoff = elected.score.to_den(1)
+		.expect("(n / d) < u128::max() and (n' / 1) == (n / d), thus n' < u128::max()'; qed.")
+		.n();
+
+	for voter_index in 0..voters.len() {
+		let voter = &mut voters[voter_index];
+
```

Comment:
I believe I indexed both `voters` and then `voters[idx].edges` by index to be able to hold mutable references to both at the same time, which is (sensibly) a violation of the borrow checking rules since `edge` is a sub-field of voter and holding a mutable reference to it implies a mutable reference to the voter as well. 

But I have to double check. So far your suggestions where all correct and my complications where not needed. I was probably confused by the compiler when I did this.